### ISSUE TITLE
Cannot update packages, because the generated -pkg.el file uses a symbol instead of a string

### ISSUE THREAD

**Issue by oantolin** [melpa/melpa#7312]
**TL;DR** (by @tarsius):

Unfortunately this bug caused the local package descriptions files (`<name>-pkg.el`) to get corrupted for everyone who updated during the last ~12 hours or so.

To fix that you have to either fix those files (1) using `find` and `sed` or (2) using elisp. Alternatively you can (3) patch `package.el` so that it accepts the corrupted files, bug fixing the files is preferable.

1. https://github.com/melpa/melpa/issues/7312#issuecomment-751192101
2. https://gist.github.com/tarsius/ee51369ee7e07821e45e123f4067a5db
3. https://github.com/melpa/melpa/issues/7312#issuecomment-751175699

---
---

```
Wrong type argument: stringp, modus-themes
```

(And not just for `modus-themes`, same things happens for any package that has just been updated.)

Looking into it, it seems the `modus-themes-pkg.el` that is generated now contains this:

```elisp
(define-package modus-themes "20201224.2015" "Highly accessible themes (WCAG AAA)"
  '((emacs "26.1"))
  :commit "cb10c65cce4b1ba2e950754d3d74c7367c088ec0" :authors
  (("Protesilaos Stavrou" . "info@protesilaos.com"))
  :maintainer
  ("Protesilaos Stavrou" . "info@protesilaos.com")
  :keywords
  ("faces" "theme" "accessibility")
  :url "https://gitlab.com/protesilaos/modus-themes")
;; Local Variables:
;; no-byte-compile: t
;; End:
```

The error is in the first line, it should have a string `"modus-themes"` instead of a symbol.

Indeed, the previous `modues-themes-pkg.el` contains this instead:

```elisp
(define-package "modus-themes" "20201223.1157" "Highly accessible themes (WCAG AAA)"
  '((emacs "26.1"))
  :commit "7b47f36571a452762e88d0a7b14deee95eb1456a" :keywords
  ("faces" "theme" "accessibility")
  :authors
  (("Protesilaos Stavrou" . "info@protesilaos.com"))
  :maintainer
  ("Protesilaos Stavrou" . "info@protesilaos.com")
  :url "https://gitlab.com/protesilaos/modus-themes")
;; Local Variables:
;; no-byte-compile: t
;; End:
```

**Comment by 4wph** [comment]
Same on my end. After attempting to install any package, the install-package command stops working (giving the same error) and afterwards emacs does not launch (also giving the same error). Only way out is removing the elpa folder altogether.

**Comment by oantolin** [comment]
> Only way out is removing the elpa folder altogether.

Or, *gasp*, using a different editor to add the quotation marks. 😉

**Comment by drgibbon** [comment]
Same for me here, Emacs 27.1.

**Comment by gusbrs** [comment]
I'm also affected by this. As other people seem to be:

https://github.com/jwiegley/use-package/issues/896

https://redd.it/kjqkwm

**Comment by alphapapa** [comment]
I think the bug may be here: https://github.com/melpa/melpa/commit/d0e82f4410e034edf3849b4fa1bb249a068a4838#diff-644b0d9504ad0eda19418af5aa12ec115b59d3ef5177e4cf9e765946f11f0525R350

**Comment by pegzmasta** [comment]
I am also experiencing this.  _dash-pkg.el_, and several other packages, uses a symbol instead of a string to define its name.  I deleted my elpa directory and tried re-installing all my packages when I came upon this error.  Now, I can't install virtually any packages, since _dash_ is one of the first few packages emacs tries to install in my setup, as it is required as a dependency by many other packages.

I tried manually editing the file to force it to be a string, but upon re-initiating emacs, _package.el_ then proceeds to break on the very next package description.  I gave up after 4 edits (seeing as I have ~100+ packages to install).  I feel like this is due to a recent change in _package.el_, because I deleted my elpa directory and re-installed all my packages successfully last month.  Could _package.el_ be **advised** to accept a string instead of a symbol (or perhaps a patch is underway)?  I do most of my computing on emacs, so... guess I'm taking a break for now :-(

**Comment by peterwang** [comment]
A dirty and quick fix could be modifying the package.el to accept both symbol and string, like this:

```diff
*** package.el.orig	2020-12-25 13:28:32.000000000 +0800
--- package.el.new	2020-12-25 13:28:32.000000000 +0800
***************
*** 419,425 ****
                  (name-string version-string &optional summary requirements
                   &rest rest-plist
                   &aux
!                  (name (intern name-string))
                   (version (version-to-list version-string))
                   (reqs (mapcar #'(lambda (elt)
                                     (list (car elt)
--- 419,425 ----
                  (name-string version-string &optional summary requirements
                   &rest rest-plist
                   &aux
! 		 (name (intern (if (stringp name-string) name-string (symbol-name name-string))))
                   (version (version-to-list version-string))
                   (reqs (mapcar #'(lambda (elt)
                                     (list (car elt)

```

**Comment by michael-heerdegen** [comment]
For me this also broke starting Emacs afterwards.  Please try to fix it fast.

**Comment by purcell** [comment]
It's literally Christmas day where I am right now but I'm having a quick look. I think @tarsius merged some changes to package-build.el from its own repo, but I haven't figured out yet whether some of those changes are a fix for this issue, and we're just not seeing it reflected in package builds yet.

**Comment by purcell** [comment]
Need to have a moratorium on merges for a couple of days around this time... :-D

**Comment by dlj-NaN** [comment]
FYI, it looks like the regression might have been introduced in `package-build--write-pkg-file` around Sep 2:

https://github.com/melpa/melpa/commit/0839bd91a4aea0a2881dc0cffd90325e3ce21b6f#diff-644b0d9504ad0eda19418af5aa12ec115b59d3ef5177e4cf9e765946f11f0525L354-L358

The current `package.el` does, in fact, seem to require quoting, contrary to the line for 56bab8e7. (0839bd91a4 was subsequently merged in 725852263)

However, `package-build--write-pkg-file` was rewritten somewhere in this squash commit:

https://github.com/melpa/melpa/commit/d0e82f4410e034edf3849b4fa1bb249a068a4838#diff-644b0d9504ad0eda19418af5aa12ec115b59d3ef5177e4cf9e765946f11f0525R350

(merged in 4a4fa64428)

**Comment by michael-heerdegen** [comment]
BTW, for all users whose Emacs fails to start due to broken package descriptions (was the case for me):

One needs to start with -q.

Then I had to edit the package descriptions of all packages I had tried
to upgrade by hand.  In ~/.emacs.d/elpa for any package directory of any
package "foo", in the directory of "foo" I needed to change the first
line of the file "foo-pkg.el":

(define-package foo ...

to

(define-package "foo" ...

to make package initialization succeed again.

**Comment by michael-heerdegen** [comment]
@purcell I dunno the code much but it isn't `package-build--write-pkg-file` using `name` (a symbol) as first arg to `define-package`?

**Comment by slondr** [comment]
> A quick and dirty fix could be modifying the package.el to accept both symbol and string package name, like this:
> 
> ```diff
> *** package.el.orig	2020-12-25 13:28:32.000000000 +0800
> --- package.el.new	2020-12-25 13:28:32.000000000 +0800
> ***************
> *** 419,425 ****
>                   (name-string version-string &optional summary requirements
>                    &rest rest-plist
>                    &aux
> !                  (name (intern name-string))
>                    (version (version-to-list version-string))
>                    (reqs (mapcar #'(lambda (elt)
>                                      (list (car elt)
> --- 419,425 ----
>                   (name-string version-string &optional summary requirements
>                    &rest rest-plist
>                    &aux
> ! 		 (name (intern (if (stringp name-string) name-string (symbol-name name-string))))
>                    (version (version-to-list version-string))
>                    (reqs (mapcar #'(lambda (elt)
>                                      (list (car elt)
> ```

@peterwang Is it possible for me to make this change on my own system? I'd like to get my emacs to boot

**Comment by lmmarsano** [comment]
@melpa Maybe this PR melpa/package-build#45 will fix it?

**Comment by io12** [comment]
> Is it possible for me to make this change on my own system? I'd like to get my emacs to boot

I did this by removing `package.elc`, and unzipping and editing `package.el.gz` in `/usr/share/emacs/27.1/lisp/emacs-lisp`. Unfortunately, I had to install Vim to do this.

**Comment by defaultxr** [comment]
[This comment](https://github.com/jwiegley/use-package/issues/896#issuecomment-751143213) suggests a one-liner shell command that, when run from your `.emacs.d` directory, fixes all of the broken files:

```
find elpa -type f -iname "*-pkg.el" -exec sed -i '/^(define-package "/! s/define-package \([a-zA-Z0-9\-]*\)/define-package "\1"/' {} \;
```

This worked for me and didn't break my computer.

**Comment by zhd173** [comment]
In macOS：
```shell
find elpa -type f -iname "*-pkg.el" -exec sed -i "" '/^(define-package "/! s/define-package \([a-zA-Z0-9\-]*\)/define-package "\1"/' {} \;
```

**Comment by purcell** [comment]
With my Christmas dinner eaten, I've now patched the cause of this, but figuring out how to force a rebuild of just the affected broken packages is work that is out of scope for this evening.

**Comment by purcell** [comment]
Looks like it might be taking care of it automatically: seems to be rebuilding everything now that package-build.el has changed. Fingers crossed the issue will have cleared itself up by my morning. Thanks all.

**Comment by robbyoconnor** [comment]
Merry Christmas @purcell and everyone else on this issue!

**Comment by tarsius** [comment]
Uh. Sorry everyone!

**Comment by a13** [comment]
>  Unfortunately, I had to install Vim to do this.

are you joking? `emacs -q` should work fine in any case

**Comment by tarsius** [comment]
@purcell Yes, next time I will just "push on Friday" instead of "push on Christmas". :roll_eyes:

**Comment by peter9207** [comment]
Thanks everyone, 

is the issue going to fix it self or do we need to take any manual action? i still get the same error as before =/

**Comment by jcs090218** [comment]
In my case, it goes away automatically. You don't need to anything.

**Comment by oantolin** [comment]
Thanks for fixing this so soon, @purcell, and on Christmas too!

**Comment by tarsius** [comment]
I think we should leave this open for a while.

**Comment by madanyang** [comment]
The oneliner and the elisp does not take into account if one is using elpa directories with versioning ie elpa-26.3 or elpa-27.1. Of course one can adjust them both manually rather then blindly copying the code. For some such a note could be helpfull

**Comment by oantolin** [comment]
> I think we should leave this open for a while.

Oh, I'm a little surprised by that. Why leave it open? Is the idea is it somehow easier to find when it's open?

At any rate, I'm very satisfied with the resolution. Please close when you feel it is appropriate.

**Comment by tarsius** [comment]
> Is the idea is it somehow easier to find when it's open?

Yes.

**Comment by tarsius** [comment]
@madanyang I've added a note.

**Comment by pegzmasta** [comment]
Thanks for the work!  No problems installing packages anymore.  Happy holidays everyone!

**Comment by purcell** [comment]
Thanks folks, I'm going to assume everything is working okay unless I hear otherwise. If a specific package still appears to be broken for anyone, post here with its name and I'll take a look.

**Comment by ivanandreev1234** [comment]
> Thanks folks, I'm going to assume everything is working okay unless I hear otherwise. If a specific package still appears to be broken for anyone, post here with its name and I'll take a look.

real-auto-save

**Comment by alphapapa** [comment]
> > Thanks folks, I'm going to assume everything is working okay unless I hear otherwise. If a specific package still appears to be broken for anyone, post here with its name and I'll take a look.
> 
> real-auto-save

What happens?

**Comment by purcell** [comment]
If it's a package you've already downloaded, it will stay broken until reinstalled or fixed using the `find` command above.
